### ISSUE TITLE
Directory.EnumerateFiles() => System.IO.DirectoryNotFoundException

### ISSUE THREAD

**Issue by diev** [dotnet/runtime#95867]
### Description

Directory.EnumerateFiles(@"D:\TEMP\James B.", "*.doc", new EnumerationOptions())
 => System.IO.DirectoryNotFoundException: "Could not find a part of the path 'D:\TEMP\James B'."

### Reproduction Steps

A full directory path ended with '.' (more long than a current directory!)

```
foreach (var file in Directory.EnumerateFiles(@"D:\TEMP\James B.", "*.doc", new EnumerationOptions()))
{
    Console.WriteLine(file);
}
```

### Expected behavior

Get file names in "D:\TEMP\James B."

### Actual behavior

```
System.IO.DirectoryNotFoundException
  HResult=0x80070003
  Message = Could not find a part of the path 'D:\TEMP\James B'.
  Source = System.Private.CoreLib
  StackTrace:
   in System.IO.Enumeration.FileSystemEnumerator`1.CreateDirectoryHandle(String path, Boolean ignoreNotFound)
   in System.IO.Enumeration.FileSystemEnumerator`1.Init()
   in System.IO.Enumeration.FileSystemEnumerable`1..ctor(String directory, FindTransform transform, EnumerationOptions options, Boolean isNormalized)
   in System.IO.Enumeration.FileSystemEnumerableFactory.UserFiles(String directory, String expression, EnumerationOptions options)
   in System.IO.Directory.InternalEnumeratePaths(String path, String searchPattern, SearchTarget searchTarget, EnumerationOptions options)
   in Program.<Main>$(String[] args) в C:\Repos\ConsoleTest\ConsoleTest\Program.cs:строка 1
```

### Known Workarounds

* string dir = "James B."; // WORKS!
* string dir = @".\James B."; // WORKS!
* string dir = Directory.GetCurrentDirectory + @"\James B."; // FAILS!

```
System.IO.DirectoryNotFoundException
  HResult=0x80070003
  Message = Could not find a part of the path 'C:\Repos\ConsoleTest\ConsoleTest\bin\Debug\net8.0\System.Func`1[System.String]\James B'.
  Source = System.Private.CoreLib
```

### Configuration

* .NET 8.0, Visual Studio 17.8.3
* Windows 11, Microsoft Windows [Version 10.0.22631.2715]
* x64

```
dotnet --info

Workload version: 8.0.100-manifests.8d38d0cc

Host:
  Version:      8.0.0
  Architecture: x64
  Commit:       5535e31a71
```

**Comment by Clockwork-Muse** [comment]
Given the following code:
```csharp
// Get the path of the temp directory
string tempPath = Path.GetTempPath();

// Create a new directory in the temp directory
string newDir = Path.Combine(tempPath, "James B.");
Directory.CreateDirectory(newDir);

// Create a new file in the new directory
string newFile = Path.Combine(newDir, "NewFile.txt");
File.WriteAllText(newFile, "Hello, world!");

// Iterate over the files in the new directory
foreach (var file in Directory.EnumerateFiles(newDir, "*.txt", new EnumerationOptions()))
{
    // Print the file name and content
    Console.WriteLine(file);
    Console.WriteLine(File.ReadAllText(file));
}
```
... cannot reproduce.
Realizing the directory path (providing the value of `newDir` as a string constant) doesn't change this.

Are you sure `D:\TEMP\` and `D:\TEMP\James B.` exist?

Do note that manually writing/combining paths with a specific separator is best avoided, and you should normally use `Path.Combine(...)`.

**Comment by diev** [comment]
You are wrong. Please look at `...\AppData\Local\Temp\James B` to see this folder is really created without that ending '.'!
Windows (i.e Explorer) cannot create a folder with that, some other programs can. After them this issue will fire!

Yes, "D:\TEMP\James B." (with a point) and "D:\TEMP\James B " (with a space) exist in my test folder as they exist in my production environment where users can create so exotic directory (and file) names. I attach a small `Test.zip` with your file where both folders I have created for you.

I use `Path.Combine()`, thanks! This example was for clearity.
[Temp.zip](https://github.com/dotnet/runtime/files/13639698/Temp.zip)

**Comment by Jozkee** [comment]
I think this is duplicate of https://github.com/dotnet/runtime/issues/88615. 

I can, however, enumerate files in `C:\Users\dacantu\AppData\Local\Temp\James B.` using dotnet's `Directory.EnumerateFiles` or `Directory.GetFiles`. 

What does not work for me is enumerate files in folders ending in space, and I'm also unable to delete them thereafter.

**Comment by diev** [comment]
> What does not work for me is enumerate files in folders ending in space, and I'm also unable to delete them thereafter.

This code healths all such mistakes:

```
    [LibraryImport("kernel32.dll", EntryPoint = "MoveFileW", SetLastError = true, StringMarshalling = StringMarshalling.Utf16)]
    [UnmanagedCallConv(CallConvs = new Type[] { typeof(System.Runtime.CompilerServices.CallConvStdcall) })]
    [return: MarshalAs(UnmanagedType.Bool)]
    private static partial bool MoveFile(string lpExistingFileName, string lpNewFileName);

    private static bool RenameDirectoryWithSideWhiteSpaces(string path, string newPath)
    {
        var oldPath = @"\\?\" + path;
        var result = MoveFile(oldPath, newPath);

        return result;
    }
```

**Comment by diev** [comment]
> I can, however, enumerate files in `C:\Users\dacantu\AppData\Local\Temp\James B.` using dotnet's `Directory.EnumerateFiles` or `Directory.GetFiles`, as @Clockwork-Muse showed in its code.

Again, are you sure this folder is CREATED as 'B.'?
Your debugger lies because your Windows has 'B' in reality.
All works until Windows/.NET tries to enumerate dirs/files inside a really named 'B.' (or 'C ') created by some other application like if you unzip my attached Test.zip

**Comment by Jozkee** [comment]
> Again, are you sure this folder is CREATED as 'B.'?

Yes, I have a folder from when I was testing https://github.com/dotnet/runtime/issues/88615 and `dir` does display it with a dot, see: 
![image](https://github.com/dotnet/runtime/assets/16040868/1ffd48af-0390-4103-94a2-f4589db0836f)

If I use minGW (git bash) you see more clearly the space in the directory:
![image](https://github.com/dotnet/runtime/assets/16040868/b215f33c-0374-4dcb-9a84-ec947d7643cf)

I just noticed that the ones I tried to create today are not keeping the dot but they keep the space, maybe it was changed in a Windows update?

I tested again and I'm able to enumerate folders ending with `.` (using `Directory.EnumerateFiles(@"\\?\C:\path.")`). Trying to enumerate the folder ending with space throws "Directory not found" error.

**Comment by diev** [comment]
@Jozkee if you remember DOS times' Norton Commander, the Windows' Far Manager (<https://farmanager.com/index.php?l=en>) is a good tool to investigate (full CRUD) such file system entries.

**Comment by diev** [comment]
The code by @Clockwork-Muse fails if an external application makes 'B.' really (Windows does not make it and the debugger lies but still works inside a function where "doesn't change this" mentioned).
But this code creates 'B.' correctly if add that undocumented (Windows only?) prefix:
```
// Get the path of the temp directory
string tempPath = @"\\?\" + Path.GetTempPath();
```
And it also read enumerated files back if that folder is really changed to 'B.' by an external application. So that magic prefix probably solves this issue for Windows if a developer remembers to add it.

**Comment by Clockwork-Muse** [comment]
> You are wrong. Please look at `...\AppData\Local\Temp\James B` to see this folder is really created without that ending '.'!

You are right, the directory created doesn't include the dot.  
Despite the fact that the combined path actually does include it.

> I can, however, enumerate files in `C:\Users\dacantu\AppData\Local\Temp\James B.` using dotnet's `Directory.EnumerateFiles` or `Directory.GetFiles`, as @Clockwork-Muse showed in its code.

I'm not, after creating the folder via WSL.
`dir` can't iterate it either, and explorer can't open it either, complaining that the dotless version doesn't exist....

So this maybe isn't a dotnet problem and maybe a Windows problem...

**Comment by diev** [comment]
> So this maybe isn't a dotnet problem and maybe a Windows problem...

As I was investigating all these problems during my long development of service utilities for actions on huge file shares on Windows servers, there is a famous case of the 260 chars limit of the oldest `MAX_PATH` constant in DOS and then Win32 API (where the real limit of NTFS is near 32K chars and Bill Gates' evangelists had advertised in 1993's: "Write now documents in their file names, not inside files!"). Old .Net Frameworks had firmly coded this constant too. My users were creating unlimited paths on servers, but Windows (still later 10!) clients and my apps on .Net Framework run on these servers could not parse them! The Win32 API had added that prefix `\\?\` to handle this case, but .Net Framework did not! Until one of the latest versions (4.6.2 in 2016?) where this firmly coded limit (as a false fool protection) was finally removed (but old Windows 10 versions still need be switched manually while Windows 8.1 Embedded are still on sale too).

So I think .NET has now two operating modes: an old Windows compatible (as Explorer ignores to create folders like 'B.') and a more modern way with the `\\?\` prefix (what of them is cross-platformed?). So that question #88615 asks why API of .NET is so unsolid.

**Comment by diev** [comment]
> ```cs
> // Create a new file in the new directory
> string newFile = Path.Combine(newDir, "New File txt"); // NOTE: NO DOTS.
> 
> // Iterate over the files in the new directory
> foreach (var file in Directory.EnumerateFiles(newDir, "*.*", new EnumerationOptions())) // NOTE: DOTS ARE.
> ```

Another probably known but rarely documented feature is that file will never be iterated until the mask will be changed to "*". Maybe here the secret of that last dot for Windows is. An attempt to create a folder name with a last '.' is like a creation of an entry with an empty extension and Windows omits it despitting your facts.

**Comment by Jozkee** [comment]
> I'm not, after creating the folder via WSL.

Can you try with extended prefix? i.e. `Path.Join("\\?\", tempPath, "James B.");`

**Comment by diev** [comment]
Currently I try to write some workarounds for my needs and I have found a new related issue:
```cs
string path = "a some path ended James B.";

string name = Path.GetFileNameWithoutExtension(path);
string ext = Path.GetExtension(path);
```
And here `name` nor `ext` has no that dot. This information becomes lost. I need to add the next code to restore it and to make a further processing on this decision:
```cs
if (path.EndsWith('.'))
    name += '.';
```
My attempts with `"\\?\"` were not very successful. So I detect problematic names and rename them through Win32 API to nearest correct names (it is required for further error-free other works). My dirty code is [here](https://github.com/diev/PropertySystem/blob/main/Diev.Extensions/Scan/PathScanner.cs). It processes all such fantasies for the moment. You may not do it because this makes disk changes instead of error-free `EnumerateFiles()`.
